File: Diagrammatize.Design.Pattern/code/cxx/Visitor/A2/Directory.cpp
#include "Directory.hpp"
#include "Visitor.hpp"
#include "SizeVisitor.hpp"

Directory::Directory(const std::string& name_) : name(name_) {}

Directory::~Directory() {}

std::string Directory::getName() {
    return name;
}

int Directory::getSize() {
    SizeVisitor v;
    accept(&v);
    return v.getSize();
}

Entry* Directory::add(Entry* entry) {
    dir.push_back(entry);
    return this;
}

void Directory::accept(Visitor* v) {
    v->visit(this);
}

File: Diagrammatize.Design.Patterns/code/python/Visitor/A2/Directory.py

from Entry import Entry
from SizeVisitor import SizeVisitor

class Directory(Entry):
    def __init__(self, name):
        self.name = name
        self.dir = []

    def getName(self):
        return self.name

    def getSize(self):
        v = SizeVisitor()
        self.accept(v)
        return v.getSize()

    def add(self, entry):
        self.dir.append(entry)
        return self

    def __iter__(self):
        return iter(self.dir)

    def accept(self, v):
        v.visitDirectory(self)
    
